### ISSUE TITLE
Selecting a category when adding a post

### ISSUE THREAD

**Issue by rowasc** [ushahidi/platform#2410]
### Intended Behavior (changes)
- [ ] A parent category should not remain selected if any children are unselected.
- [ ] Everyone should be able to add pre-existing categories to a post if they have the permissions for the individual categories and permissions to edit the post.
- [ ] Users with permissions should be able to link a category to a post
- [ ] If there are no children in the parent category, we DO let the user select the parent/top-level category when adding a post. 
- [ ] If there are children in the parent category, we DON'T let the user select the parent/top-level category when adding a post. 

#### "Everyone can see" permission works correctly when linking a category to a post 
- [x] Open the "Add new post"  screen as an admin. 
- [x] Fill out required info
- [x] Select a pre-existing category with "Everyone can see" permission and save.
- [x] Select only that category for this post. 
- [x] Save the post (draft/in-review)
- [x] Select only the newly created child category as a filter
- [x] You should see the post you just linked to that category in the list
- [x] You should see other posts linked to that category in the list 
- [x] Publish the post
- [x] Logout
- [x] Select only the category you just used as a filter. 
- [x] You should see the post you just linked to that category in the list
- [x] You should see other published posts linked to that category in the list 

#### Parent categories are disabled/can not be selected if they have children 

- [x] Open the "Add new post"  screen as an admin. 
- [x] Fill out required info
- [x] Try to select a parent category (that has children) 
- [x] You should not be able to select the parent category

#### Parent categories can be selected if they don't have children 

- [x] Open the "Add new post"  screen as an admin. 
- [x] Fill out required info
- [x] Try to select a parent category (that does not have children) 
- [x] You should be able to select the parent category

#### Parent categories are public by default when created from the Add Post or Edit Post screen
- [x] Open the "Add new post"  screen as an admin. 
- [x] Fill out required info
- [x] Create a new category.
- [x] Select only that category for this post. 
- [x] Save the post (draft/in-review)
- [x] Select only the newly created category as a filter
- [x] You should see the post you just linked to that category in the list 
- [x] You should see no other posts since you only linked the category to the post you created it with.
- [x] Publish the post
- [x] Logout
- [x] Select only the category you just used as a filter. 
- [x] You should see the post you just created with that category in the list
- [x] You should see no other posts since you only linked the category to the post you created it with.

#### Parent categories are disabled/cannot be selected if they have children  in the survey creation/survey edition screens when adding a category field
- [ ] Open the "Add new post"  screen as an admin. 
- [ ] Fill out required info
- [ ] Try to select a parent category (that has children) 
- [ ] You should NOT be able to select the parent category

#### When I only select 1 child out of many in the Survey Editor for a category field, I only see that child and its parent when I open the "Add new post" screen for that type.
Screenshots as example:
survey editor screen
<img width="592" alt="screen shot 2018-02-19 at 12 40 02 pm" src="https://user-images.githubusercontent.com/2434401/36385715-1825ce64-1572-11e8-944c-3e7778a7f29a.png">

add new post screen
<img width="547" alt="screen shot 2018-02-19 at 12 40 09 pm" src="https://user-images.githubusercontent.com/2434401/36385713-17fb17aa-1572-11e8-9954-56e3bc2560c6.png">

- [ ] Open the "Survey editor" for a survey, and add a category field with only 1 child selected for a parent category
- [ ] Fill out required info, and save. 
- [ ] Open the "add new post" screen
- [ ] Verify that you do not see other child categories besides the one you made available for that category field

------------------------------ 
------ REMOVED DURING PHASE . This test case should not be here. Does not apply since we cannot create child categories in add new post ------ 
#### Child categories use parent permissions by default when created from the Add Post or Edit Post screen
- [x] Open the "Add new post"  screen as an admin. 
- [x] Fill out required info
- [ ] Create a new category under a parent (make sure the parent category can only be seen as admin).
- [ ] Select only the newly created child category for this post. 
- [ ] Save the post (draft/in-review)
- [ ] Select only the newly created child category as a filter
- [ ] You should see the post you just linked to that category in the list
- [ ] You should see no other posts since you only linked the category to the post you created it with.
- [ ] Publish the post
- [ ] Logout
- [ ] In the filters drop-down, you should not see the parent or the children category since they are admin-only. 
------------------------------ 

### Why it is important to clarify this and decide what goes into the dev phase for export

If we don't understand this behavior, we don't know what we are exporting. If we don't understand it, a user won't, either. CSV Exports should be reliable and filters working plays a big role.  Not all the parts of this are equally relevant to filtering.

**Comment by rowasc** [comment]
@aoduor we are following the permissions conversation here. I think this could be dropped in this cycle, since it's not a blocker for CSV (Inconsistent permissions ARE a blocker, but specifically who can add categories or not, doesn't sound like a blocker.
@justinscherer @Erioldoesdesign 

@Angamanga thoughts?

**Comment by aoduor** [comment]
> Which users have permissions to add categories to a post?
This should be everyone, subject to individual category permissions. 

> Who has permission to add a category on the fly?
Limit this to admin, OR, add it as a permission to be assigned when creating roles. 

> What are default permissions of a category added on the fly?
I'm not sure what the best option here would be, and would love to hear other people's opinion/for us to test this out. My initial thought is leave it as public

**Comment by rowasc** [comment]
Thanks @aoduor 
Added your answers. I agree.
@Erioldoesdesign can you take a look at the issue, and let me know if this is reasonable or if you feel we need to do a session or chat about it?

**Comment by rowasc** [comment]
responding in more detail 
@aoduor 
>Which users have permissions to add categories to a post?
>> Angela: This should be everyone, subject to individual category permissions.

Agree. 
Sadly, the add category widget is kind of terrible in the add post, you can't even add child categories in there. I don't know if that's something we want to fix now (if it would block export from being stable) to be honest. I think it has to be a separate issue in the product backlog. 


>Who has permission to add a category on the fly?
>> Angela: Limit this to admin, OR, add it as a permission to be assigned when creating roles.

I think for this cycle we can use the same permission check we have when creating a category from settings? hasManageSettingsPermission which is assigned to admin-type users normally. But we can create a lower priority issue to make it available to everyone? 

> Q: What are default permissions of a category added on the fly?
>> Angela: I'm not sure what the best option here would be, and would love to hear other people's opinion/for us to test this out. My initial thought is leave it as public


I think for parents, it's reasonable for it to be using the "Visible to everyone" setting, by default while adding it in the post creation page. Plus, this is the current behavior so it's not a change we need to make. 
For children we don't have a way to add child categories in the Add new post screen :( So I'm pushing it to a separate ticket in the product backlog. 


LMK If this sounds reasonable and its clear, as a starting point for CSV Export to be usable.
I can create the tickets when confirmed.

**Comment by Angamanga** [comment]
Looks good, just one question and added a part that might be ok to move to another ticket.

#### Question
> ### Child categories use parent permissions by default when created from the Add Post or Edit Post screen
> - [ ] In the filters dropdown, you should see the parent category but not the one you just created as an admin
I dont agree with this behaviour, I dont think the parent-category should be visible either?

#### Added part
> Open the "add new post" as not logged in
> Select a survey with a category-field but without any categories with "Everyone can see" Permissions
> The category-field should not be seen at all

Actually, this could be moved to another ticket but I think its deeply annoying and kind of embarrassing that you get Only a part that says "Categories" and the box with "Select all" And no categories if there are no categories that you have the permissions to see in that survey...
![screen shot 2018-02-01 at 10 05 11](https://user-images.githubusercontent.com/8624777/35723865-17365b3c-07fc-11e8-9473-b34804851f64.png)

**Comment by Angamanga** [comment]
Moved ui-changes to this issue: https://app.zenhub.com/workspace/o/ushahidi/platform/issues/2455

**Comment by crcommons** [comment]
Seeing some confusion amongst many issues:
*A parent category should not remain selected if all children are unselected.*

This should read:

_A parent category should not remain selected if ANY children are unselected._

**Comment by rowasc** [comment]
@crcommons @Angamanga edited the intended behavior. THat was an oversight. Thanks

**Comment by rowasc** [comment]
@crcommons @Angamanga corrected a few typos and added a mention to set the filters where it was missing (an omission, the intent was the same). 

Should be OK now

**Comment by rowasc** [comment]
Sending to spec until @crcommons or @Angamanga review

**Comment by rowasc** [comment]
Hold the review. Waiting on a confirmation from Eriol and Angie on the following: 

- YOU CAN select just a parent for a post. You should be able to. Because you want to be able to assign to a general “parent” category without selecting children. Something might be categorized as a fruit, but you don’t need to know and add every fruit that exists for it to be a fruit. 
- YOU CANNOT select just a parent when filtering.

**Comment by aoduor** [comment]
> A parent category should not remain selected if ANY children are unselected.

![screenshot](https://media.giphy.com/media/3oFzmte1x5WVErIqFq/giphy.gif)

This should apply for both submission and filtering.

**Comment by Erioldoesdesign** [comment]
_YOU CAN select just a parent for a post. You should be able to. Because you want to be able to assign to a general “parent” category without selecting children. Something might be categorized as a fruit, but you don’t need to know and add every fruit that exists for it to be a fruit._

Yes, I think this is still useful to users, being able to assign 'Fruit' to something that you have a good idea is a fruit, but your not exactly sure which type of fruit. The ideal behaviour is that everyone would be able to look at an apple and say 'That's an apple' but not everyone will know that and may just say 'Well at least I know it's a fruit'.
I suspect there's an improvement here where posts can be edited/re-categorised when people (admins perhaps?) know in more detail where this is best situated. (This may exist already)

_YOU CANNOT select just a parent when filtering._
Agreed - selecting 'Fruit' parent selects all subsequent 'mangoes, apples...' children fruit. 

The only other potential demystifying change to this for users is that we treat the parent as 'unspecified' in relation to children. But again, best tackled when we do a category rehaul.

**Comment by aoduor** [comment]
> YOU CAN select just a parent for a post. You should be able to. Because you want to be able to assign to a general “parent” category without selecting children. Something might be categorized as a fruit, but you don’t need to know and add every fruit that exists for it to be a fruit.

NO. Especially during submission, the parent is a heading. You should not be able to assign it without selecting children, unless it is childless/standalone. Unless anyone else has a different opinion from this. Users can only select from the options presented to them( ie pick the fruits we have on the menu, not the ones we don't have)

>YOU CANNOT select just a parent when filtering.
You CAN select a parent alone when filtering( as long as all it's children are unselected. If you have some children selected, the parent should not be selected. 

i.e
1. When filtering, I may want to know/see all fruits regardless of their specificity = SELECT PARENT ONLY. because it encompasses all.
2. The moment i select to see mangoes, and apples, i'm not looking at all fruits, i'm being specific, and we have declared independence from the mother tag, fruits. = DESELECT PARENT, select children.

**Comment by aoduor** [comment]
@Erioldoesdesign see ^^

**Comment by Erioldoesdesign** [comment]
@aoduor Definitely pro-people being more specific when selecting categories. Also happy to defer here because this is a very involved ticket about a subject that I have less understanding of at present.

**Comment by rowasc** [comment]
@aoduor @Erioldoesdesign ugh. I'm sorry. It's the description that was up for review (since we chatted in slack to get clarity after I posted those comments). Basically, yes I agree with this 💯 , I think the description agrees with the comment Angie sent since I wrote it based on our chat. ?
Thanks

**Comment by rowasc** [comment]
Resolved. This is spec'd correctly.

**Comment by rowasc** [comment]
@Angamanga @jshorland demo here https://drive.google.com/open?id=1iEqtwF74epYBsi_Pq_5Bfe1_h7U0itHS
Note - important: I made a mistake in the last test scenario I added. That scenario "Child categories use parent permissions by default when created from the Add Post or Edit Post screen" does not make sense as a test case for this ticket because we don't have a way to add children categories (only parents) from add post/edit post. I "removed" (/made a note saying it's invalid) on that test case since it does not help with filtering and it's a `new feature`, not a fix relevant to csv filtering. I think we should add a new ticket for design so we have a way to do this in the future (add children from add post screen), agree? 
Bonus =) => I ended up fixing an old bug that wasn't noted here that made it hard to test these scenarios, essentially in the past, we were not updating the category list when a new category was added in the Add Post or Edit post screen, and it was pretty awful/hard to test without fixing it.

**Comment by jshorland** [comment]
WOO! nice double up on that bug fix. That's been around since uchaguzi.

Yes to ticket for creating children categories from within the survey form

**Comment by rowasc** [comment]
Excellent. 
Creating ticket today

**Comment by rowasc** [comment]
@aoduor added this  test case since it affected the same code: 
#### Parent categories are NOT disabled/can be selected if they have children  in the `survey creation/survey edition screens when adding a category field`
- [x] Open the "Add new post"  screen as an admin. 
- [x] Fill out required info
- [x] Try to select a parent category (that has children) 
- [x] You should be able to select the parent category

Is that OK? It'd work the same as before when adding a category field basically

**Comment by rowasc** [comment]
@Angamanga see https://github.com/ushahidi/platform/issues/2410#issuecomment-365344379

**Comment by jshorland** [comment]
## TESTING Script
#### "Everyone can see" permission works correctly when linking a category to a post 
- [] Open the "Add new post"  screen as an admin. 
- [] Fill out required info
- [] Select a pre-existing category with "Everyone can see" permission and save.
- [] Select only that category for this post. 
- [] Save the post (draft/in-review)
- [] Select only the newly created child category as a filter
- [] You should see the post you just linked to that category in the list
- [] You should see other posts linked to that category in the list 
- [] Publish the post
- [] Logout
- [] Select only the category you just used as a filter. 
- [] You should see the post you just linked to that category in the list
- [] You should see other published posts linked to that category in the list 

#### Parent categories are disabled/can not be selected if they have children 

- [] Open the "Add new post"  screen as an admin. 
- [] Fill out required info
- [] Try to select a parent category (that has children) 
- [] You should not be able to select the parent category

#### Parent categories can be selected if they don't have children 

- [] Open the "Add new post"  screen as an admin. 
- [] Fill out required info
- [] Try to select a parent category (that does not have children) 
- [] You should be able to select the parent category

#### Parent categories are public by default when created from the Add Post or Edit Post screen
- [] Open the "Add new post"  screen as an admin. 
- [] Fill out required info
- [] Create a new category.
- [] Select only that category for this post. 
- [] Save the post (draft/in-review)
- [] Select only the newly created category as a filter
- [] You should see the post you just linked to that category in the list 
- [] You should see no other posts since you only linked the category to the post you created it with.
- [] Publish the post
- [] Logout
- [] Select only the category you just used as a filter. 
- [] You should see the post you just created with that category in the list
- [] You should see no other posts since you only linked the category to the post you created it with.

## Addition to TESTING Script (devphase)
#### Parent categories are NOT disabled/can be selected if they have children  in the survey creation/survey edition screens when adding a category field
- [] Open the "Add new post"  screen as an admin. 
- [] Fill out required info
- [] Try to select a parent category (that has children) 
- [] You should be able to select the parent category

**Comment by jshorland** [comment]
A few things:

1. what happens when I create a child to a category AFTER i've already created posts tagged with ONLY the parent category?
eg: https://cycle-zero-dev.lab2.ushahidilabs.com/posts/1
https://cycle-zero-dev.lab2.ushahidilabs.com/posts/2

2. Select all has very weird behavior when submitting a post. When I click any category, it selects all, and when I click select all again, it turns them all off, but then clicking select all doesn't do anything. See gif. 

![Screen Capture on 2018-02-14 at 14-28-46.gif](https://images.zenhubusercontent.com/5a81e5e34b5806bc2bc6a08f/698cd2bc-e333-4177-9f03-1809198e695a)

3. I cannot select the parent category at all. It gives me a strikethrough icon indicating "no" when I move my mouse to it. is this correct?
https://d.pr/i/cpdpFm (it won't let me upload the gif)

**Comment by rowasc** [comment]
going one by one: 
1. I think that's the correct behavior. Tagging @aoduor . 
@aoduor see:
>>> what happens when I create a child to a category AFTER i've already created posts tagged with ONLY the parent category?
eg: https://cycle-zero-dev.lab2.ushahidilabs.com/posts/1
https://cycle-zero-dev.lab2.ushahidilabs.com/posts/2


2. UH ok that's a bug bug bug. I'll fix that ASAP 

3. Yea that's the expected behavior for #2410, you cannot select a parent directly if it has children already

**Comment by rowasc** [comment]
Fix before EOC : when adding a category field in the survey editor, the parent categories should look to the user as in the Add Post scenario. The only difference is that when the user selects a child, the parent is also sent in the edit/create request . So if you have
- Parent 
    - Child 1
    - Child 2
And Select Child 2, your category "Parent" will also be associated to that category field. This is necessary since we use the structure later when adding posts and filtering. The benefit of not letting children select parents only (unless they don't have children) is that we enforce the same behavior in all the screens. 
Discussed & Confirmed this with Angie last friday .

**Comment by rowasc** [comment]
Added the 2 new test cases in the description (devphase addition)

**Comment by rowasc** [comment]
@jshorland two demos. 
Fix for select-all and fix the add post vs add field scenario
https://drive.google.com/open?id=1ByLEYB-iDwb0kliJcPHQbhPJCreHePBu
https://drive.google.com/open?id=1drPLd2pQNxph6b8M9uQm1jBXNIpL9VRC

**Comment by jshorland** [comment]
looks good

**Comment by rowasc** [comment]
@rowasc to deploy this in a testing env
